### PR TITLE
Addition of NDO module for IPSLA Monitoring Policy (DCNE-135)

### PR DESCRIPTION
**Pull request by shrsr**

### CODE REVIEW

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        template=dict(type="str", required=True),
+        ipsla_monitoring_policy=dict(type="str", aliases=["name"]),
+        ipsla_monitoring_policy_uuid=dict(type="str", aliases=["uuid"]),
+        sla_type=dict(type="str", choices=["http", "tcp", "icmp", "l2ping"], default="icmp"),
+        destination_port=dict(type="int", default=0),
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:183]: do we need defaults for this? since typically we do not define defaults.

File: tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+      register: nm_create_new_ipsla_monitoring_policy_again
+
+    - name: Assert IPSLA monitoring policy was created
+      assert:
+        that:
+          - cm_create_new_ipsla_monitoring_policy is changed
+          - cm_create_new_ipsla_monitoring_policy.previous == {}
+          - cm_create_new_ipsla_monitoring_policy.current.name == "ansible_test_ipsla_monitoring_policy"
+          - cm_create_new_ipsla_monitoring_policy.current.slaType == "icmp"
+          - cm_create_new_ipsla_monitoring_policy.current.slaFrequency == 60
```
Comment (akinross) [inline tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml:117]: if defaults should we assert those here?

**Review by akinross — CHANGES_REQUESTED** [review]

File: tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml (hunk 2 above)
Comment (shrsr) [reply tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml:118]: Removed the defaults

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 1 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py]: Removed them

**Review by sajagana — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    KVPair,
+)
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        template=dict(type="str", required=True),
+        ipsla_monitoring_policy=dict(type="str", aliases=["name"]),
+        ipsla_monitoring_policy_uuid=dict(type="str", aliases=["uuid"]),
```
[the remark this answers is not in the document]
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:176]: Done!

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+    http_version: HTTP11
+    http_uri: /test
+    sla_frequency: 120
+    detect_multiplier: 5
+    req_data_size: 64
+    type_of_service: 16
+    operation_timeout: 1000
+    threshold: 1000
+    ipv6_traffic_class: 32
+    state: present
```
Comment (gmicol) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:124]: Wouldn't it better for this example to make the sla_type to TCP and therefore add the destination_port attribute? Like that, we have a full example with all attributes being used.

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+    type: str
+    aliases: [ uuid ]
+  sla_type:
+    description:
+    - The type of SLA.
+    type: str
+    choices: [ http, tcp, icmp, l2ping ]
+  destination_port:
+    description:
+    - The port used for SLA.
```
Comment (gmicol) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:56]:
```suggestion
    - The port used for SLA.
    - Only used when O(sla_type=tcp) otherwise it will be ignored.
```
Something like that.

**Review by gmicol — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 4 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:124]: If we use TCP http_version, http_uri are not required.

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by gmicol — APPROVED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+  ipsla_monitoring_policy_uuid:
+    description:
+    - The uuid of the IPSLA Monitoring Policy.
+    - This parameter is required when the O(ipsla_monitoring_policy) needs to be updated.
+    type: str
+    aliases: [ uuid ]
+  sla_type:
+    description:
+    - The type of SLA.
+    - This is only applicable when O(sla_type=tcp).
```
Comment (anvitha-jain) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:52]: This is only applicable when O(sla_type=tcp).  has to be a part of destination_port, after line 57

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+
+- name: Delete an IPSLA monitoring policy
+  cisco.mso.ndo_ipsla_monitoring_policy:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    template: ansible_tenant_template
+    ipsla_monitoring_policy: ansible_test_ipsla_monitoring_policy
+    state: absent
+"""
```
Comment (anvitha-jain) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:159]: can you example for query with UUID and delete with UUID

File: tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+      register: query_all
+    
+    - name: Assert IPSLA monitoring policy was queried
+      assert:
+        that:
+          - query_one is not changed
+          - query_one.current.name == "ansible_test_ipsla_monitoring_policy_2"
+          - query_all is not changed
+          - query_all.current | length == 2
+
```
Comment (anvitha-jain) [inline tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml:301]: test case for query with UUID is missing

File: tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+          - nm_create_another_ipsla_monitoring_policy.current.httpUri == "/test"
+          - nm_create_another_ipsla_monitoring_policy.current.slaFrequency == 120
+          - nm_create_another_ipsla_monitoring_policy.current.slaDetectMultiplier == 5
+          - nm_create_another_ipsla_monitoring_policy.current.reqDataSize == 64
+          - nm_create_another_ipsla_monitoring_policy.current.ipv4ToS == 16
+          - nm_create_another_ipsla_monitoring_policy.current.timeout == 1000
+          - nm_create_another_ipsla_monitoring_policy.current.threshold == 1000
+          - nm_create_another_ipsla_monitoring_policy.current.ipv6TrfClass == 32
+
+    - name: Delete the IPSLA monitoring policy
```
Comment (anvitha-jain) [inline tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml:474]: Test case for delete with UUID is missing

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

**Review by anvitha-jain — DISMISSED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+ANSIBLE_METADATA = {
+    "metadata_version": "1.1",
+    "status": ["preview"],
+    "supported_by": "community",
+}
+
+DOCUMENTATION = r"""
+---
+module: ndo_ipsla_monitoring_policy
+short_description: Manage IPSLA Monitoring Policies on Cisco Nexus Dashboard Orchestrator (NDO).
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:21]: Shall we write IP SLA with a space in between since this is how it is typically documented in other docs?

```suggestion
short_description: Manage IP SLA Monitoring Policies on Cisco Nexus Dashboard Orchestrator (NDO).
```

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+    description:
+    - The type of SLA.
+    type: str
+    choices: [ http, tcp, icmp, l2ping ]
+  destination_port:
+    description:
+    - The port used for SLA.
+    - This is only applicable when O(sla_type=tcp).
+    type: int
+  http_version:
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:59]: Should we add `This is only applicable when O(sla_type=http)` to the description of those arguments that are specific to a sla type?

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+    aliases: [ uuid ]
+  sla_type:
+    description:
+    - The type of SLA.
+    type: str
+    choices: [ http, tcp, icmp, l2ping ]
+  destination_port:
+    description:
+    - The port used for SLA.
+    - This is only applicable when O(sla_type=tcp).
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:57]: This is not entirely true I think since http would default to 80 if not mistaken?

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+    - The value must be in the range 0 - 255.
+    type: int
+  operation_timeout:
+    description:
+    - The operation_timeout for SLA.
+    type: int
+  threshold:
+    description:
+    - The threshold for SLA.
+    type: int
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:95]: perhaps add the metric? milliseconds?
```suggestion
  operation_timeout:
    description:
    - The operation_timeout for SLA in milliseconds.
    type: int
  threshold:
    description:
    - The threshold for SLA in milliseconds.
    type: int
```

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+    choices: [ HTTP10, HTTP11 ]
+  http_uri:
+    description:
+    - The URI used for HTTP SLA.
+    type: str
+  sla_frequency:
+    description:
+    - The frequency of SLA monitoring.
+    - The value must be in the range 1 - 300.
+    type: int
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:72]:
```suggestion
  sla_frequency:
    description:
    - The frequency of SLA monitoring in seconds.
    - The value must be in the range 1 - 300.
    type: int
```

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+  detect_multiplier:
+    description:
+    - The detection multiplier for SLA.
+    - The value must be in the range 1 - 100.
+    type: int
+  req_data_size:
+    description:
+    - The size of the request data.
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:80]:
```suggestion
    - The size of the request data in bytes.
```

File: tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,518 @@
...
+          - nm_create_another_ipsla_monitoring_policy.current.slaPort == 80
+          - nm_create_another_ipsla_monitoring_policy.current.httpVersion == "HTTP11"
+          - nm_create_another_ipsla_monitoring_policy.current.httpUri == "/test"
+          - nm_create_another_ipsla_monitoring_policy.current.slaFrequency == 120
+          - nm_create_another_ipsla_monitoring_policy.current.slaDetectMultiplier == 5
+          - nm_create_another_ipsla_monitoring_policy.current.reqDataSize == 64
+          - nm_create_another_ipsla_monitoring_policy.current.ipv4ToS == 16
+          - nm_create_another_ipsla_monitoring_policy.current.timeout == 1000
+          - nm_create_another_ipsla_monitoring_policy.current.threshold == 1000
+          - nm_create_another_ipsla_monitoring_policy.current.ipv6TrfClass == 32
```
Comment (akinross) [inline tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml:482]: Is this needed? You are already creating a second one for query all, cant you use that one for the uuid delete test or is there another reason for this test?

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        template=dict(type="str", required=True),
+        ipsla_monitoring_policy=dict(type="str", aliases=["name"]),
+        description=dict(type="str"),
+        ipsla_monitoring_policy_uuid=dict(type="str", aliases=["uuid"]),
+        sla_type=dict(type="str", choices=["http", "tcp", "icmp", "l2ping"]),
+        destination_port=dict(type="int"),
+        http_version=dict(type="str", choices=["HTTP10", "HTTP11"]),
+        http_uri=dict(type="str"),
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:205]: So we need to error when type is http and these are not provided or do they default to a value when not provided? If they default we should document this. Else we likely need some additional input validation required_by or something.

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        template=dict(type="str", required=True),
+        ipsla_monitoring_policy=dict(type="str", aliases=["name"]),
+        description=dict(type="str"),
+        ipsla_monitoring_policy_uuid=dict(type="str", aliases=["uuid"]),
+        sla_type=dict(type="str", choices=["http", "tcp", "icmp", "l2ping"]),
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:202]: Do we need any specific type / input validation? Like icmp type and http version for instance makes no sense to provide.

**Review by akinross — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 12 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:57]: That's true, I'll add that here.

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 17 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:205]: These optional http attributes default to a value. I have updated the docs.

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 18 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:202]: The API actually accepts values when icmp is used with the optional http type attributes. But I'll go ahead and add a validation.

File: tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml (hunk 16 above)
Comment (shrsr) [reply tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml:482]: This is mainly for code coverage and not for deletion via uuid. Without this some part of code is not covered under inclusion of attributes within payload when type is http. In the above cases we add update to http from icmp.

Comment (akinross) [reply tests/integration/targets/ndo_ipsla_monitoring_policy/tasks/main.yml:482]: Perhaps rename the test case to make it clear that this is what is tested?

**Review by gmicol — APPROVED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 19]
```diff
@@ -0,0 +1,495 @@
...
+    type: int
+  threshold:
+    description:
+    - The threshold for SLA in milliseconds.
+    type: int
+  ipv6_traffic_class:
+    description:
+    - The IPv6 Traffic Class.
+    - The value must be in the range 0 - 255.
+    type: int
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:105]: do these default to some value? it seems so from the tests, should we document?

**Review by akinross — CHANGES_REQUESTED** [review]

**Review by akinross — APPROVED** [review]
LGTM

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

**Review by samiib — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 20]
```diff
@@ -0,0 +1,502 @@
...
+ANSIBLE_METADATA = {
+    "metadata_version": "1.1",
+    "status": ["preview"],
+    "supported_by": "community",
+}
+
+DOCUMENTATION = r"""
+---
+module: ndo_ipsla_monitoring_policy
+short_description: Manage IP SLA Monitoring Policies on Cisco Nexus Dashboard Orchestrator (NDO).
```
[the remark this answers is not in the document]
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:21]: Decided to go with IPSLA

**Review by akinross — APPROVED** [review]
LGTM

**Review by gmicol — APPROVED** [review]
LGTM

**Review by samiib — APPROVED** [review]
LGTM

**Review by anvitha-jain — DISMISSED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 21]
```diff
@@ -0,0 +1,502 @@
...
+    - The value must be in the range 1 - 300.
+    - The default value is 60 seconds.
+    type: int
+  detect_multiplier:
+    description:
+    - The detection multiplier for SLA.
+    - The value must be in the range 1 - 100.
+    - The default value is 3.
+    type: int
+  req_data_size:
```
Comment (lhercot) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:85]:
```suggestion
  request_data_size:
```

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 22]
```diff
@@ -0,0 +1,502 @@
...
+  sla_type:
+    description:
+    - The type of SLA.
+    type: str
+    choices: [ http, tcp, icmp, l2ping ]
+  destination_port:
+    description:
+    - The port used for SLA.
+    - This is only applicable when O(sla_type=tcp) or O(sla_type=http).
+    - The value defaults to 80 when O(sla_type=http).
```
Comment (lhercot) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:58]: What happens when it is tcp?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 22 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:58]: Changed the description

**Review by anvitha-jain — DISMISSED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 23]
```diff
@@ -0,0 +1,503 @@
...
+    - The value must be in the range 1 - 65535 when O(sla_type=tcp).
```
[the remark this answers is not in the document]
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:59]: As mentioned above the value defaults to 80 when the type is http

**Review by gmicol — APPROVED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (samiib) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:60]: I can see it defaults to 80 for `http` but the description says the port range should be 1 - 65535 when `tcp`.
It doesn't appear to mention a valid port range for `http`.
So basically the description implies that when type is `tcp` a port must be provided in the range of 1 - 65535.
However, when the type is `http` the port defaults to 80 and I presume there is also a range validation between 1 - 65535?

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 24]
```diff
@@ -0,0 +1,504 @@
...
+  ipsla_monitoring_policy_uuid:
+    description:
+    - The UUID of the IPSLA Monitoring Policy.
+    - This parameter is required when the O(ipsla_monitoring_policy) needs to be updated.
+    type: str
+    aliases: [ uuid ]
+  sla_type:
+    description:
+    - The type of SLA.
+    - - The default value is icmp.
```
Comment (samiib) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:52]:
```suggestion
    - The default value is icmp.
```

**Review by samiib — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:60]: By the description "The value defaults to 80 when O(sla_type=http)" I am stating that the module automatically sets it to 80 when user specifies the sla_type as http. There is no validation range for http here and http is commonly associated with port 80. Let me know if you have suggestions to change the description to state this clearer?

**Review by akinross — APPROVED** [review]
LGTM

**Review by samiib — DISMISSED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (samiib) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:60]: Sure, makes sense if there is no range validation for http.

**Review by gmicol — DISMISSED** [review]
LGTM

**Review by anvitha-jain — DISMISSED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 25]
```diff
@@ -0,0 +1,504 @@
...
+    - The port used for SLA.
+    - This is only applicable when O(sla_type=tcp) or O(sla_type=http).
+    - The value defaults to 80 when O(sla_type=http).
+    - The value must be in the range 1 - 65535 when O(sla_type=tcp).
+    type: int
+  http_version:
+    description:
+    - The HTTP version used for SLA.
+    - This is only applicable when O(sla_type=http).
+    - The default value is HTTP10.
```
Comment (lhercot) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:66]:
```suggestion
    - The default value is C(HTTP10).
```

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 26]
```diff
@@ -0,0 +1,504 @@
...
+  ipsla_monitoring_policy_uuid:
+    description:
+    - The UUID of the IPSLA Monitoring Policy.
+    - This parameter is required when the O(ipsla_monitoring_policy) needs to be updated.
+    type: str
+    aliases: [ uuid ]
+  sla_type:
+    description:
+    - The type of SLA.
+    - The default value is icmp.
```
Comment (lhercot) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:52]:
```suggestion
    - The default value is C(icmp).
```

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by anvitha-jain — APPROVED** [review]
LGTM

**Review by samiib — DISMISSED** [review]
LGTM

**Review by akinross — APPROVED** [review]
LGTM

**Review by gmicol — APPROVED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 27]
```diff
@@ -0,0 +1,504 @@
...
+    type: str
+    aliases: [ name ]
+  description:
+    description:
+    - The description of the IPSLA Monitoring Policy.
+    type: str
+  ipsla_monitoring_policy_uuid:
+    description:
+    - The UUID of the IPSLA Monitoring Policy.
+    - This parameter is required when the O(ipsla_monitoring_policy) needs to be updated.
```
Comment (lhercot) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:46]:
```suggestion
    - This parameter is required when the O(ipsla_monitoring_policy) attribute needs to be updated.
```

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by gmicol — DISMISSED** [review]
LGTM

**Review by samiib — DISMISSED** [review]
LGTM

**Review by akinross — APPROVED** [review]
LGTM

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 28]
```diff
@@ -0,0 +1,504 @@
...
+        )
+        if match:
+            mso.existing = mso.previous = copy.deepcopy(match.details)
+    else:
+        mso.existing = mso.previous = existing_ipsla_policies
+
+    if state == "present":
+        if match:
+            if ipsla_monitoring_policy and match.details.get("name") != ipsla_monitoring_policy:
+                ops.append(
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:313]: @sajagana  wrote a function for this, asked him to do a separate PR for it so we can rebase the open PRs on it. Shoudl we do this?
https://github.com/CiscoDevNet/ansible-mso/pull/563/files#diff-f3e3720de5c754cfeb714aa1ed0f9d0a1a043f8c0005944d58565decefe1817bR31

Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:313]: Does this help in any way?

Comment (samiib) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:313]: I think the function is helpful to reduce complexity and code duplication.
However, maybe we can start using once its merged in instead of having to rebase everything?

**Review by anvitha-jain — APPROVED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 28 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py]: Done!

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 29]
```diff
@@ -0,0 +1,398 @@
...
+
+            if threshold:
+                payload["threshold"] = threshold
+
+            if ipv6_traffic_class:
+                payload["ipv6TrfClass"] = ipv6_traffic_class
+
+            ops.append(dict(op="add", path="{0}/-".format(path), value=payload))
+
+            mso.sanitize(payload)
```
Comment (samiib) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:373]: You can also move the `replace_data` outside the if statement and use it here. The `sanitize` function with `collate=True` will remove any unset keys and set `mso.sent` to the payload.

```suggestion
            mso.sanitize(replace_data, collate=True)
            ops.append(dict(op="add", path="{0}/-".format(path), value=mso.sent))
```

**Review by samiib — CHANGES_REQUESTED** [review]

**Review by sajagana — APPROVED** [review]
LGTM!

File: plugins/modules/ndo_ipsla_monitoring_policy.py [hunk 30]
```diff
@@ -0,0 +1,359 @@
...
+            existing_ipsla_policies,
+            [KVPair("uuid", ipsla_monitoring_policy_uuid) if ipsla_monitoring_policy_uuid else KVPair("name", ipsla_monitoring_policy)],
+        )
+        if match:
+            mso.existing = mso.previous = copy.deepcopy(match.details)
+    else:
+        mso.existing = mso.previous = existing_ipsla_policies
+
+    if state == "present":
+        replace_data = {
```
Comment (akinross) [inline plugins/modules/ndo_ipsla_monitoring_policy.py:312]: should we use dict() or {}? we use dict() most of the time for dict instantiation, but not sure if it matters

Comment (akinross) [reply plugins/modules/ndo_ipsla_monitoring_policy.py:312]: samita names it mso_values, which is the variable name I followed also now, should we align on that?https://github.com/CiscoDevNet/ansible-mso/pull/574/files#diff-56139042a23bb757d511b1f82886f1ce58db3f27bc1942c0fc4831fa1c7a4223R262

**Review by akinross — CHANGES_REQUESTED** [review]

File: plugins/modules/ndo_ipsla_monitoring_policy.py (hunk 30 above)
Comment (shrsr) [reply plugins/modules/ndo_ipsla_monitoring_policy.py]: Done

**Review by samiib — DISMISSED** [review]
LGTM

**Review by gmicol — DISMISSED** [review]
LGTM

**Review by samiib — APPROVED** [review]
LGTM

**Review by sajagana — APPROVED** [review]
LGTM!

**Review by gmicol — APPROVED** [review]
LGTM

**Review by akinross — APPROVED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]

**Review by lhercot — APPROVED** [review]
LGTM
